packages/react-dom/src/server/ReactPartialRenderer.js

```diff
@@ -483,6 +489,12 @@ class ReactDOMServerRenderer {
 
   constructor(children: mixed, makeStaticMarkup: boolean) {
     var flatChildren;
+    if (
+      React.isValidElement(children) &&
```

Comment:
If this is false then we're doing this check twice now. Maybe restructure so we don't have to?

Reply:
In order to preserve the optimization on the if/else, I'd end up with something like:
```
    if (React.isValidElement(children)) {
      const element = ((children: any): ReactElement);
      if (element.type === REACT_FRAGMENT_TYPE) {
        if (React.isValidElement(element.props.children)) {
          flatChildren = [element.props.children];
        } else {
          flatChildren = toArray(element.props.children);
        }
      } else {
        flatChildren = toArray(element);
      }
    } else {
      flatChildren = toArray(children);
    }
```
The code is a bit more confusing, but it does save a duplicate check. What do you think?

Reply:
Can you extract this in a function with early returns?

```js

    function flattenTopLevelChildren(children: mixed): FlatReactChildren {
      if (!React.isValidElement(children)) {
        return toArray(children);
      }
      const element = ((children: any): ReactElement);
      if (element.type !== REACT_FRAGMENT_TYPE) {
        return [element];
      }
      const fragmentChildren = element.props.children;
      if (!React.isValidElement(fragmentChildren)) {
        return toArray(fragmentChildren);
      }
      const fragmentChildElement = ((fragmentChildren: any): ReactElement);
      return [fragmentChildElement];
    }
```

Reply:
This looks very nice. 👍